consensus/types/src/blob_sidecar.rs

```diff
@@ -1,17 +1,46 @@
-use crate::test_utils::TestRandom;
-use crate::{Blob, ChainSpec, Domain, EthSpec, Fork, Hash256, SignedBlobSidecar, SignedRoot, Slot};
-use bls::SecretKey;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::sync::Arc;
+
 use derivative::Derivative;
 use kzg::{Kzg, KzgCommitment, KzgPreset, KzgProof};
 use rand::Rng;
+use serde::de::DeserializeOwned;
 use serde_derive::{Deserialize, Serialize};
-use ssz::Encode;
+use ssz::{Decode, Encode};
 use ssz_derive::{Decode, Encode};
 use ssz_types::{FixedVector, VariableList};
-use std::sync::Arc;
-use test_random_derive::TestRandom;
+use tree_hash::TreeHash;
 use tree_hash_derive::TreeHash;
 
+use bls::SecretKey;
+use test_random_derive::TestRandom;
+
+use crate::test_utils::TestRandom;
+use crate::{Blob, ChainSpec, Domain, EthSpec, Fork, Hash256, SignedBlobSidecar, SignedRoot, Slot};
+
+pub enum SidecarListVariant<E: EthSpec> {
```

Comment:
We should be able to get rid of both `SidecarListVariant` and `BlobsOrBlobRoots` by relying more heavily on `AbstractExecPayload` and `Sidecar`. Adding a new associated type to represent `BlobsOrBlobRoots` should work. Or you could maybe add it as an associated type to the `Sidecar` trait. Not sure which would be simplest. And you'll probably need to add a trait for this associated type so you can work with blinded and non-blinded versions. 

I'm not sure this would make it prettier, but the blinded vs full variants of the enums should always be consistent with `AbstractExecPayload` so we can guarantee this by using generics instead of enums and I think it would make it clearer what parts of a type are generic over the same thing

Reply:
Thanks for the review! 🙏  I tried to get rid of it and struggle a bit but I'll give it another try! Might be able to get it working now after reading your network refactoring PR ☺️

Reply:
I hacked together a version of this and it's messy and incomplete but compiling if you want to use it as inspiration https://github.com/realbigsean/lighthouse/tree/moar-generics

The naming's bad and the structure could probably be improved but it might help you with some parts

Reply:
Nice!! That's a good cleanup actually - I attempted pretty much the same thing (adding an associated type to `Sidecar`) but ran into issues with generic types conversion!

I think these trait functions here are they key you used to resolved this! Awesome, thanks I'll get this change onto this branch!

```rust
pub trait RawBlobs<T: EthSpec>: Sync + Send {
    fn from_blob_roots(roots: BlobRoots<T>) -> Self;
    fn from_blobs(blobs: Blobs<T>) -> Self;
    fn len(&self) -> usize;
    fn is_empty(&self) -> bool;
}
```

Reply:
I'm working through some clean ups - will push a new version tomorrow

Reply:
I've cherry-picked your commits and cleaned up the todos, please see diffs: 97ee1f49d97c990dc21c06b742198afd2362cadc.